### PR TITLE
[Merged by Bors] - feat(Order/Max): basic `IsMin` / `IsBot` lemmas

### PR DESCRIPTION
**Pull request by vihdzp**
---

I couldn't find anything similar to these, but there's a possibility they might already be in Mathlib somewhere.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Order/Max.lean [hunk 1]
```diff
@@ -332,8 +332,48 @@ protected theorem IsMax.eq_of_le (ha : IsMax a) (h : a ≤ b) : a = b :=
...
+protected theorem IsTop.not_isBot [Nontrivial α] (ha : IsTop a) : ¬ IsBot a :=
+  fun ha' ↦ ha.toDual.not_isTop ha'.toDual
+
 end PartialOrder
 
+section LinearOrder
+
+variable [LinearOrder α] {a : α}
+
+protected theorem IsMin.isBot (ha : IsMin a) : IsBot a := by
```
[the remark this answers is not in the document]
Comment (vihdzp) [reply Mathlib/Order/Max.lean:355]: Nice catch! Yeah, `IsMin.isBot` already exists under that setting.

[the remark this answers is not in the document]
Comment (vihdzp) [reply Mathlib/Order/Max.lean]: What about `IsMin.not_isMax`? Those seem to work under either a directed or a codirected order, is that the correct generality?

[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/Order/Max.lean]: Think so yeah

File: Mathlib/Order/Directed.lean [hunk 2]
```diff
@@ -250,18 +250,64 @@ theorem isBot_iff_isMin [IsDirected α (· ≥ ·)] : IsBot a ↔ IsMin a :=
...
+section PartialOrder
+
+variable [PartialOrder β]
 
-theorem exists_lt_of_directed_ge [IsDirected β (· ≥ ·)] [Nontrivial β] : ∃ a b : β, a < b := by
+section Nontrivial
+
+variable [Nontrivial β]
+
+theorem exists_lt_of_directed_ge (β) [PartialOrder β] [IsDirected β (· ≥ ·)] [Nontrivial β] :
```
Comment (urkud) [inline Mathlib/Order/Directed.lean:281]: What changed here? Why do you have to repeat all variables?

Comment (vihdzp) [reply Mathlib/Order/Directed.lean:281]: The variable `β` was accidentally being made explicit for all subsequent lemmas.

File: Mathlib/Order/Max.lean [hunk 3]
```diff
@@ -332,6 +332,20 @@ protected theorem IsMax.eq_of_le (ha : IsMax a) (h : a ≤ b) : a = b :=
 protected theorem IsMax.eq_of_ge (ha : IsMax a) (h : a ≤ b) : b = a :=
   h.antisymm' <| ha h
 
+protected theorem IsBot.lt_of_ne (ha : IsBot a) (h : a ≠ b) : a < b :=
+  (ha b).lt_of_ne h
+
+protected theorem IsTop.lt_of_ne (ha : IsTop a) (h : b ≠ a) : b < a :=
+  (ha b).lt_of_ne h
+
+protected theorem IsBot.not_isTop [Nontrivial α] (ha : IsBot a) : ¬ IsTop a := by
```
Comment (urkud) [inline Mathlib/Order/Max.lean:341]: You can prove `IsBot.not_isMax`, then deduce this using `mt IsTop.isMax`.

File: Mathlib/Order/Directed.lean (hunk 2 above)
Comment (urkud) [reply Mathlib/Order/Directed.lean:281]: You can use `variable (β) in` before each lemma where it should be explicit (or use `variable (β)`, then `variable {β}`). This way it's clear that you reuse TC assumptions from `variable`s, just change `β` to explicit.

**Comment by urkud** [comment]
Thanks! Delegating to @YaelDillies 
bors d=@YaelDillies

File: Mathlib/Order/Directed.lean [hunk 4]
```diff
@@ -250,18 +250,66 @@ theorem isBot_iff_isMin [IsDirected α (· ≥ ·)] : IsBot a ↔ IsMin a :=
...
+    IsDirected α (· ≥ ·) :=
+  ⟨fun a b => ⟨a ⊓ b, inf_le_left, inf_le_right⟩⟩
+
+-- see Note [lower instance priority]
+instance (priority := 100) OrderTop.to_isDirected_le [LE α] [OrderTop α] : IsDirected α (· ≤ ·) :=
+  ⟨fun _ _ => ⟨⊤, le_top _, le_top _⟩⟩
+
+-- see Note [lower instance priority]
+instance (priority := 100) OrderBot.to_isDirected_ge [LE α] [OrderBot α] : IsDirected α (· ≥ ·) :=
+  ⟨fun _ _ => ⟨⊥, bot_le _, bot_le _⟩⟩
```
Comment (YaelDillies) [inline Mathlib/Order/Directed.lean:271]: Can you put this back to the end of the file? The first two are about `Semilattice...` which is stronger than `PartialOrder`, and I don't see why you need to move the last two?

Comment (vihdzp) [reply Mathlib/Order/Directed.lean:271]: Yeah, I don't remember why I moved these. I'll move them back.

**Comment by YaelDillies** [comment]
maintainer merge

**Comment by YaelDillies** [comment]
bors merge

Silly me

**Pull request closed without merging**